cli/sign.go

```diff
@@ -0,0 +1,203 @@
+	signFlags.StringVar(&InfoLocation, "location", "", "Location of the signatory")
+	signFlags.StringVar(&InfoReason, "reason", "", "Reason for signing")
+	signFlags.StringVar(&InfoContact, "contact", "", "Contact information for signatory")
+	signFlags.StringVar(&TSA, "tsa", "https://freetsa.org/tsr", "URL for Time-Stamp Authority")
+	signFlags.StringVar(&CertType, "certType", "CertificationSignature", "Type of the certificate (CertificationSignature, ApprovalSignature, UsageRightsSignature, TimeStampSignature)")
+
+	signFlags.Usage = func() {
+		fmt.Printf("Usage: %s sign [options] <input.pdf> <output.pdf> <certificate.crt> <private_key.key> [chain.crt]\n\n", os.Args[0])
+		fmt.Println("Sign a PDF file with a digital signature")
+		fmt.Println("\nOptions:")
+		signFlags.PrintDefaults()
+		fmt.Println("\nExamples:")
+		fmt.Printf("  %s sign -name \"John Doe\" input.pdf output.pdf cert.crt key.key\n", os.Args[0])
+		fmt.Printf("  %s sign -certType \"TimeStampSignature\" input.pdf output.pdf\n", os.Args[0])
+	}
+
+	if err := signFlags.Parse(os.Args[2:]); err != nil {
+		log.Fatalf("Failed to parse sign flags: %v", err)
+	}
+
+	if len(signFlags.Args()) < 1 {
+		signFlags.Usage()
+		os.Exit(1)
+	}
+
+	input := signFlags.Arg(0)
+	SignPDF(input, signFlags.Args())
+}
+
+func SignPDF(input string, args []string) {
+	certTypeValue, err := ParseCertType(CertType)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if certTypeValue == sign.TimeStampSignature {
+		if len(args) < 2 {
+			fmt.Fprintf(os.Stderr, "TimeStamp signing requires: input.pdf output.pdf\n")
+			os.Exit(1)
+		}
+		output := args[1]
+		TimeStampPDF(input, output, TSA)
+		return
+	}
+
+	if len(args) < 4 {
+		fmt.Fprintf(os.Stderr, "Signing requires: input.pdf output.pdf certificate.crt private_key.key [chain.crt]\n")
+		os.Exit(1)
+	}
+
+	output := args[1]
+	certPath := args[2]
+	keyPath := args[3]
+	var chainPath string
+	if len(args) > 4 {
+		chainPath = args[4]
+	}
+
+	cert, pkey, certificateChains := LoadCertificatesAndKey(certPath, keyPath, chainPath)
+
+	err = sign.SignFile(input, output, sign.SignData{
+		Signature: sign.SignDataSignature{
+			Info: sign.SignDataSignatureInfo{
+				Name:        InfoName,
+				Location:    InfoLocation,
+				Reason:      InfoReason,
+				ContactInfo: InfoContact,
+				Date:        time.Now().Local(),
+			},
+			CertType:   certTypeValue,
+			DocMDPPerm: sign.AllowFillingExistingFormFieldsAndSignaturesPerms,
+		},
+		Signer:            pkey,
+		DigestAlgorithm:   crypto.SHA256,
+		Certificate:       cert,
+		CertificateChains: certificateChains,
+		TSA: sign.TSA{
+			URL: TSA,
+		},
+	})
+	if err != nil {
+		log.Println(err)
+	} else {
+		log.Println("Signed PDF written to " + output)
+	}
+}
+
+func LoadCertificatesAndKey(certPath, keyPath, chainPath string) (*x509.Certificate, crypto.Signer, [][]*x509.Certificate) {
+	certData, err := os.ReadFile(certPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	certBlock, _ := pem.Decode(certData)
```

Comment:
If certBlock is nil and we have rest bytes this might be a der encoded certificate. We could provide that directly to parse certificate.